File: c++/Solutions/largestSumAfterKNegations.cpp
#include <algorithm>
#include <numeric>
#include <cmath>

#include "largestSumAfterKNegations.hpp"


namespace largestSumAfterKNegations
{
    int Solution::largestSumAfterKNegations(std::vector<int>& A, int K)
    {
        if (A.size() == 0)
        {
            return 0;
        }
        if (A.size() == 1)
        {
            if (K%2 == 0)
            {
                return A[0];
            }
            return 0-A[0];
        }
        std::sort(A.begin(),A.end());
        auto nagativeNumber = getNegativeNumbers(A);
        if (nagativeNumber > K)
        {
            auto partNegative = std::accumulate(A.begin(),A.begin() + K, 0);
            auto partrest = std::accumulate(A.begin() + K, A.end(), 0);
            return partrest - partNegative;
        }

        auto unNegationPart = std::accumulate(A.begin() + nagativeNumber, A.end(), 0);
        auto negationPart = std::accumulate(A.begin(),A.begin() + nagativeNumber, 0);

        if ((K -nagativeNumber)  % 2 == 0)
        {
            return unNegationPart - negationPart;
        }

        int balanceFactor {0};
        if (nagativeNumber == 0)
        {
            balanceFactor = A[nagativeNumber];
        }
        else
        {
            if (nagativeNumber == A.size())
            {
                balanceFactor = abs(A[nagativeNumber-1]);
            }
            else
            {
                balanceFactor = std::min(abs(A[nagativeNumber-1]), A[nagativeNumber]);
            }
        }
        return unNegationPart - negationPart - 2 * balanceFactor;
    };

    int Solution::getNegativeNumbers(std::vector<int>& A)
    {
        auto firstNative = std::find_if(A.begin(), A.end(), [](int num){return num >= 0;});
        if (firstNative == A.end())
        {
            return A.size();
        }
        return std::distance(A.begin(),firstNative);
    }
}

File: python/solution/largestSumAfterKNegations.py
from typing import List

class Solution:
    def largestSumAfterKNegations(self, A: List[int], K: int) -> int:
        if len(A) == 0:
            return 0
        if len(A) == 1:
            if K % 2 == 1:
                return -A[0]
            else:
                return A[0]
        
        increase_list = A.copy()
        increase_list.sort()
        negativeNumbers = self.__getNegativeNumvers(increase_list)
        if negativeNumbers > K:
            partNegative = sum(increase_list[0:K])
            partrest = sum(increase_list[K:])
            return partrest - partNegative

        unNegationPart = sum(increase_list[negativeNumbers:])
        negationPart = sum(increase_list[0:negativeNumbers])

        if (K-negativeNumbers) % 2 == 0:
            return unNegationPart - negationPart

        balanceFactor = 0
        if negativeNumbers == 0:
            balanceFactor = increase_list[0]
        if negativeNumbers == len(increase_list):
            balanceFactor = 0 - increase_list[negativeNumbers-1]
        else:
            balanceFactor = min(abs(increase_list[negativeNumbers-1]),increase_list[negativeNumbers])
        
        return unNegationPart - negationPart - 2 * balanceFactor



    def __getNegativeNumvers(self, A: List[int]) -> int:
        for i,item in enumerate(A):
            if item >= 0:
                if i == 0:
                    return 0
                else:
                    return i
        return len(A)

listin = [-4,-6,9,-2,2]
solution = Solution()

res = solution.largestSumAfterKNegations(listin,4)
print(res)